### PR TITLE
4583 custom suite name

### PR DESCRIPTION
**Pull request by otemek**
My proposal of implementation for suite names.
Any questions, ideas, or remarks are welcome as it's my first contribution to the RF project.

### CODE REVIEW

**Review by pekkaklarck — CHANGES_REQUESTED** [review]
On a quick look this looks really promising. I need to take a bit closer look at the code changes and added tests later, but I already noticed these things that need to be addressed:

- The setting name should be just `Name`, not `Suite Name`, to be consistent with `Documentation` and `Metadata` that are suite specific as well. It won't then be consistent with `Suite Setup/Teardown`, but they need the `Suite` prefix to separate them from `Test Setup/Teardown`.
- The setting needs to be translateable. This requires adding `name_setting = None` to the `Language` base class in the `robot.conf.languages` module and English translation to the `En` class. You can add Finnish translation `Nimi` there as well.
- This functionality needs to be documented in the User Guide. This can be done separately after merging this PR as well.
-

**Comment by otemek** [comment]
Thanks for the help and I hope I understood it correctly because I've  left the token name SUITE_NAME and changed only the value for it as there is already a token "NAME" for tokenizing e.g. keyword names (am I right?)

**Comment by pekkaklarck** [comment]
I mainly meant the name of the setting you use in the data. I noticed you already changed it to `Name` which is great.

What token name/type to use is a good question. It didn't occur to me there's already `NAME` token that's apparently used with settings. I'd in general prefer having less token types and would be fine reusing `NAME` also with suite names, but we already have `TESTCASE_NAME` and `KEYWORD_NAME` so probably it's best to add `SUITE_NAME` as well. The value of the token type shouldn't really matter, but we've in general used values that match the name like
```python
    TESTCASE_NAME = 'TESTCASE NAME'
    KEYWORD_NAME = 'KEYWORD NAME'
```
so I guess this should be
```python
    SUITE_NAME = 'SUITE NAME'
```

**Review by pekkaklarck — CHANGES_REQUESTED** [review]
The added tests look pretty comprehensive, but unfortunately they aren't structured correctly. As the [atest/README.rst](https://github.com/robotframework/robotframework/tree/master/atest#readme) file explains, tests on `atest/robot` side should just validate that features tested on `atest/testdata` side work correctly. Basically all files using the new `Name` setting should be on the `atest/testdata` side and on the `atest/robot` side you'd have tests that validate that the name is set correctly.

There's existing [atest/robot/parsing/suite_settings.robot](https://github.com/robotframework/robotframework/blob/master/atest/robot/parsing/suite_settings.robot) file for testing other settings. It runs [atest/testdata/parsing/suite_settings.robot](https://github.com/robotframework/robotframework/blob/master/atest/testdata/parsing/suite_settings.robot) and validates everything works correctly. It currently checks that `${SUITE.name}` is set correctly based on the file name, but you could add `Name` setting to the file and see that the name is set accordingly. A problem with this approach is that it would be good to have an explicit test for getting suite name from a file name as well. In addition to that, it's a good idea to test that `Name` works also with `__init__.robot` files and that with directories the name is got from the directory name by default.

An alternative would be having a separate `atest/robot/parsing/suite_name.robot` file that tests all aforementioned combinations. Also it would need to run some tests suites where `Name` is used (and not used). Instead of creating new dummy suites, it would probably be better to add `Name` to few suites under [atest/testdata/misc/suites](https://github.com/robotframework/robotframework/tree/master/atest/testdata/misc/suites) and run it. Also other tests run `atest/testdata/misc/suites` so you may need to update them after making changes, but that shouldn't be too big a task.

I didn't yet look at the code in detail but on a very quick look it looks really good. I'll review it in more detail once tests have been restructured. If you have any questions about the somewhat complicated test setup or anything else, don't hesitate to ask here or on the #devel channel on our Slack.

File: atest/testdata/parsing/suite_names/__init__.robot
Comment (pekkaklarck) [inline atest/testdata/parsing/suite_names/__init__.robot:4]: This setting should be just `Name`, not `Suite Name`.

No need to use tags here. If there was a need, should use `Test Tags` instead of deprecated `Force Tags`. Finally, we don't generally use empty lines between settings and following the same style would keep tests consistent.

File: atest/testdata/parsing/suite_names/child_tests01/child0102.robot
Comment (pekkaklarck) [inline atest/testdata/parsing/suite_names/child_tests01/child0102.robot:10]: Also this test is unnecessary. Removing it and just leaving `Verify Suite Name` makes the test simpler. Same comment applies to latter suites with similar structure.

The empty line between settings is unnecessary as well.

File: src/robot/conf/languages.py
Comment (pekkaklarck) [inline src/robot/conf/languages.py:312]: It's interesting that the setting has name `Name` here but it still needs to be `Suite Name` in data. Most likely this is due to lexer having `Suite Name`. To make sure the used name matches, existing translation tests should also use `Name` setting. It ought to be enough to use `Nimi` (the Finnish translation) in `atest/testdata/parsing/translations/finnish/tests.robot` and enhance `Validate Translations` keyword in `atest/robot/parsing/translations.robot` so that it validates `${SUITE.name}` similarly as e.g. `${SUITE.doc}`.

File: src/robot/parsing/lexer/tokens.py
Comment (pekkaklarck) [inline src/robot/parsing/lexer/tokens.py:52]: It's a bit unfortunate we apparently need to use `SUITE_NAME` here instead of `NAME` that's already used elsewhere.

File: utest/parsing/test_lexer.py
Comment (pekkaklarck) [inline utest/parsing/test_lexer.py:46]: Change to `Name` also in unit tests.

**Review by pekkaklarck — CHANGES_REQUESTED** [review]
Code looks good but some things still to be done:
1. There are now tests only on the `atest/testdata` side which means they aren't actually run as part of the normal acceptance test run. Separate `atest/robot/parsing/suite_names.robot` file is needed to run them and to validate that suite names are set correctly. If you need help with this, ping me on Slack or ask help here.
2. As commented separately, it would be good to test also translation.
3. The code uses `Suite Name` in some places and `Name` in others. Should consistently use `Name` everywhere.
4. Documentation in the User Guide is missing. That can be done separately once this PR is otherwise ready.

**Comment by otemek** [comment]
Hi, thanks @pekkaklarck for extensive review, but i haven't finished re-work after your previous review. I know I shouldn't commit unfinished job, but I was in the middle of doing it (analyze what's exactly is going on there...) and than had to switch to another task... nevertheless, I'll try to complete this thing on the nearest weekend

File: atest/testdata/parsing/suite_names/__init__.robot
Comment (otemek) [inline atest/testdata/parsing/suite_names/__init__.robot:1]: to be deleted as for that already existing suite will be used

File: atest/testdata/parsing/suite_names/child_tests01/__init__.robot
Comment (otemek) [inline atest/testdata/parsing/suite_names/child_tests01/__init__.robot:1]: to be deleted as for testing this feature - already existing suite will be used

File: atest/testdata/parsing/suite_names/child_tests01/child0101.robot
Comment (otemek) [inline atest/testdata/parsing/suite_names/child_tests01/child0101.robot:1]: to be deleted as for testing this feature - already existing suite will be used

File: atest/testdata/parsing/suite_names/child_tests01/child0102.robot
Comment (otemek) [inline atest/testdata/parsing/suite_names/child_tests01/child0102.robot:1]: to be deleted as for testing this feature - already existing suite will be used

File: atest/testdata/parsing/suite_names/child_tests02/child0201.robot
Comment (otemek) [inline atest/testdata/parsing/suite_names/child_tests02/child0201.robot:1]: to be deleted as for testing this feature - already existing suite will be used

File: atest/testdata/parsing/suite_names/child_tests02/child0202.robot
Comment (otemek) [inline atest/testdata/parsing/suite_names/child_tests02/child0202.robot:1]: to be deleted as for testing this feature - already existing suite will be used

File: atest/testdata/parsing/suite_names/suite_w_name.robot
Comment (otemek) [inline atest/testdata/parsing/suite_names/suite_w_name.robot:1]: to be deleted as for testing this feature - already existing suite will be used

File: atest/testdata/parsing/suite_names/suite_wo_name.robot
Comment (otemek) [inline atest/testdata/parsing/suite_names/suite_wo_name.robot:1]: to be deleted as for testing this feature - already existing suite will be used

File: atest/robot/core/filter_by_names.robot
Comment (pekkaklarck) [inline atest/robot/core/filter_by_names.robot:4]: Please remove. We don't generally use tags like this. We do use tags for other purposes, though, and unnecessary tags like this are just a distraction. If tags like this were used, the value should be aligned to match other setting values.

File: atest/robot/parsing/suite_names.robot
Comment (pekkaklarck) [inline atest/robot/parsing/suite_names.robot:5]: 1. Please remove the commented Test Tags line. No need to have such tags and we avoid commented out code as well.
2. Align setting values.

File: atest/robot/parsing/suite_names.robot
Comment (pekkaklarck) [inline atest/robot/parsing/suite_names.robot:23]: 1. What does "Multi Root" mean here?
2. Add newline at the end of the file.

File: atest/testdata/misc/suites/subsuites2/__init__.robot
Comment (pekkaklarck) [inline atest/testdata/misc/suites/subsuites2/__init__.robot:2]: 1. It would probably be a good idea to include the word "Custom" in all suite names to make it explicit they are custom names. You used that style with most other suite names.
2. Add newline at the end of the file.

These comments apply also to other suites.

File: src/robot/conf/languages.py
Comment (pekkaklarck) [inline src/robot/conf/languages.py:310]: For consistency single quotes would be better than double quotes. Same also with other `name_settings`.

File: src/robot/parsing/lexer/settings.py
Comment (pekkaklarck) [inline src/robot/parsing/lexer/settings.py:114]: - Nice way to handle the fact that the setting name and type don't match.
- For consistency it would be good to use either single or double quotes in the dictionary. We've generally used single quotes (and ought to even mention that in contribution guidelines).

**Review by pekkaklarck — CHANGES_REQUESTED** [review]
Awesome work! I think all my comments were about style. Two other things to do in addition to them:

1. Add localization test. You could just add e.g. `# language: Polish` into some of the data files and use use `Nazwa` there instead of `Name`. Possibly existing localization tests could also be enhanced, but I can do that separately afterwards as well.
2. Add documentation to the User Guide. I can take care of this as well.

I hope you can take a look my comments and possibly also the two other points above. If you are busy, I can also merge this PR and make these changes myself.

File: src/robot/parsing/lexer/settings.py
Comment (otemek) [reply src/robot/parsing/lexer/settings.py:114]: switched to single quote

File: src/robot/conf/languages.py
Comment (otemek) [reply src/robot/conf/languages.py:310]: switched to single quotes

File: atest/testdata/misc/suites/subsuites2/__init__.robot
Comment (otemek) [reply atest/testdata/misc/suites/subsuites2/__init__.robot:2]: changed name to `Custom parent Suite Name`
+ new EOL

File: atest/robot/core/filter_by_names.robot
Comment (otemek) [reply atest/robot/core/filter_by_names.robot:4]: test tags removed. Just forgot to remove them as I'm using those to debug only those particular tests using "--include" switch

File: atest/robot/parsing/suite_names.robot
Comment (otemek) [reply atest/robot/parsing/suite_names.robot:5]: tag removed
settings aligned

File: atest/robot/parsing/suite_names.robot
Comment (otemek) [reply atest/robot/parsing/suite_names.robot:9]: - correct variable name to use uppercase as the global variable
- changed `Should be Equal As Strings` to `Should be Equal`
- removed word `Simple` from test case names

File: atest/robot/parsing/suite_names.robot
Comment (otemek) [reply atest/robot/parsing/suite_names.robot:23]: Changed name from `Multi Root' to 'Multi Init' to indicate that the suite name is placed under multiple folders with `__init__.robot` initialization file with setting `Name`. Can it be like that?

File: utest/parsing/test_lexer.py
Comment (otemek) [reply utest/parsing/test_lexer.py]: Changed to `Name`

File: src/robot/parsing/model/statements.py
Comment (otemek) [reply src/robot/parsing/model/statements.py]: Changed to `Name`

File: src/robot/parsing/lexer/settings.py
Comment (otemek) [reply src/robot/parsing/lexer/settings.py]: changed to `Name`

File: src/robot/conf/languages.py
Comment (otemek) [reply src/robot/conf/languages.py]: moved before `documentation_setting`

File: src/robot/conf/languages.py
Comment (otemek) [reply src/robot/conf/languages.py]: Changed to `Name` in  data

File: src/robot/parsing/lexer/tokens.py
Comment (otemek) [reply src/robot/parsing/lexer/tokens.py:52]: Probably not the best time to ask, but after going through this PR one more time I feel like I'm confused about which Token should i use... either reuse `Token.NAME` or the new one `Token.SUITE_NAME` :smile:

File: atest/robot/parsing/suite_names.robot
Comment (pekkaklarck) [reply atest/robot/parsing/suite_names.robot:23]: I don't understand what you mean with "placed under multiple folders". Is the file inside a sub folder that's executed? If yes, then something like `Custom child suite name` would be better. Anyway, the exact name doesn't matter that much.

File: atest/robot/core/filter_by_names.robot
Comment (pekkaklarck) [inline atest/robot/core/filter_by_names.robot:4]: Why wasn't this removed?

File: atest/robot/parsing/suite_names.robot
Comment (pekkaklarck) [inline atest/robot/parsing/suite_names.robot:5]: Why wasn't this removed?

**Comment by pekkaklarck** [comment]
I enabled running tests on CI and several tests failed due to changes to suite names under `atest/testdata/misc`. That's as expected but these tests obviously need to be fixed. Can you @otemek take care of that?

It also seems that although my comments have been marked as resolved, changes aren't seen here. Did you forget to commit and push changes?

Other than the above this PR is ready to be merged. Still need to write documentation, but I already promised to do that earlier.

File: atest/robot/parsing/suite_names.robot
Comment (otemek) [reply atest/robot/parsing/suite_names.robot]: renamed to `Custom Child Suite Name`

**Comment by otemek** [comment]
Sorry about those delays/code not pushed/lags/etc... but as You probably have noticed - I have quite a big problem with finding some time for me to completely take care of this so I don't know if I'll be able to finish this in a reasonable time. If there is anyone else who can make it quicker then I'm totally fine with passing it to someone else :)
P.S. Sorry for the problems

**Comment by pekkaklarck** [comment]
No worries! We all have our time pressures and with some new features there's surprisingly lot to do to get everything done. If you don't have time to continue, that's totally fine. We can either continue from here or post-pone this to RF 6.2.

One reason there are so many tests to be fixed is that you added a new test under `atest/testdata/misc/suites`. That breaks all tests running that directory and checking test counts. I'm not sure is the new test needed in this case, though. I believe it would have been possible to just change suite names which wouldn't have had such a big effect on other tests. Unless that new test is really needed, I'd just remove the whole `subsubsuites3`.

**Comment by pekkaklarck** [comment]
I looked at the changes again and this is _so_ close to be ready that I believe we should get this in still in RF 6.1.

As I wrote above, the new test under `atest/testdata/misc/suites` is the reason why so many existing tests need to be changed. You seem to have already fixed many of them and I now enabled CI again to see are there still failures. If there are, removing the new test may be easier way forward than fixing the remaining failures. Some changes would need to be reverted after removing the test but that is likely a smaller task.

**Comment by pekkaklarck** [comment]
I try to get those tests fixed and fixes pushed to this same PR!

**Comment by pekkaklarck** [comment]
I fixed test and pushed changes here. Looking at the history, there are now some commits that are related to recent changes in master, not to actual changes in this PR. I don't know are they due to me pushing these changes without first merging/rebasing locally or were they here already. Well, if tests pass I'll merge this PR and squash all changes. Hopefully that makes the history clean as well.

**Comment by pekkaklarck** [comment]
Rebasing the PR branch locally fixed the history but then needed to force-push changes here. Changes look good and I'll merge this if the CI run succeeds.

**Comment by otemek** [comment]
Thanks for the help and for making it done

**Comment by pekkaklarck** [comment]
Thanks for your contribution!

**Pull request merged**
